### PR TITLE
test(eslint-plugin): migrate validation tools to jest test cases

### PR DESCRIPTION
**Pull request by armano2**
This is follow up PR that changes way how we do test configs and documentations.

There is few positives and negatives with changing it to tests:
- Custom tools are more descriptive and can/show custom messages like "run x command"
- Tests uses common framework to validate if configurations/documents are valid, it means that they can be simpler, because we can utilize jest functions (like comparing objects), that allows us to reduce complexity.

For example new configs validation is going to look like this
```diff
all.json config › contains all @typescript-eslint/eslint-plugin rule modules, except the deprecated ones

expect(received).toEqual(expected) // deep equality
- Expected
+ Received

@@ -2,10 +2,11 @@
   "@typescript-eslint/adjacent-overload-signatures": "error",
   "@typescript-eslint/array-type": "error",
   "@typescript-eslint/await-thenable": "error",
   "@typescript-eslint/ban-ts-ignore": "error",
   "@typescript-eslint/ban-types": "error",
+  "@typescript-eslint/brace-style": "error",
   "@typescript-eslint/camelcase": "error",
   "@typescript-eslint/class-name-casing": "error",
   "@typescript-eslint/consistent-type-assertions": "error",
   "@typescript-eslint/consistent-type-definitions": "error",
   "@typescript-eslint/explicit-function-return-type": "error",
```

TODO:
- [x] update test tiles

https://github.com/typescript-eslint/typescript-eslint/pull/1396#issuecomment-570118743

### CODE REVIEW

File: packages/eslint-plugin/package.json [hunk 1]
```diff
@@ -30,8 +30,8 @@
   "main": "dist/index.js",
   "scripts": {
     "build": "tsc -b tsconfig.build.json",
-    "check:docs": "../../node_modules/.bin/ts-node --files --transpile-only ./tools/validate-docs/index.ts",
-    "check:configs": "../../node_modules/.bin/ts-node --files --transpile-only ./tools/validate-configs/index.ts",
+    "check:docs": "jest tests/docs.test.ts --runTestsByPath --silent --runInBand",
+    "check:configs": "jest ests/configs.test.ts --runTestsByPath --silent --runInBand",
```
Comment (armano2) [inline packages/eslint-plugin/package.json]:
```suggestion
    "check:configs": "jest tests/configs.test.ts --runTestsByPath --silent --runInBand",
```

File: packages/eslint-plugin/tests/configs.test.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      ([, rule]) =>
+        rule.meta.docs.recommended !== false &&
+        rule.meta.docs.requiresTypeChecking !== true,
+    )
+    .map<[string, string]>(([name, rule]) => [
+      `${RULE_NAME_PREFIX}${name}`,
+      rule.meta.docs.recommended || 'off',
+    ]);
+
+  it('contains all @typescript-eslint/eslint-plugin rule modules, except the deprecated ones', () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/configs.test.ts]:
```suggestion
  it('contains all recommended rules that don't require typechecking, excluding the deprecated ones', () => {
```

File: packages/eslint-plugin/tests/configs.test.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      ([, rule]) =>
+        rule.meta.docs.recommended !== false &&
+        rule.meta.docs.requiresTypeChecking === true,
+    )
+    .map<[string, string]>(([name, rule]) => [
+      `${RULE_NAME_PREFIX}${name}`,
+      rule.meta.docs.recommended || 'off',
+    ]);
+
+  it('contains all @typescript-eslint/eslint-plugin rule modules, except the deprecated ones', () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/configs.test.ts]:
```suggestion
  it('contains all recommended rules that require type checking, excluding the deprecated ones', () => {
```

File: packages/eslint-plugin/tests/configs.test.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+const RULE_NAME_PREFIX = '@typescript-eslint/';
+
+describe('all.json config', () => {
+  const configRules = filterRules(plugin.configs.all.rules);
+  const ruleConfigs = notDeprecatedRules.map<[string, string]>(([name]) => [
+    `${RULE_NAME_PREFIX}${name}`,
+    'error',
+  ]);
+
+  it('contains all @typescript-eslint/eslint-plugin rule modules, except the deprecated ones', () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/configs.test.ts]:
```suggestion
  it('contains all of the rules, excluding the deprecated ones', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  );
+  const readme = marked.lexer(readmeRaw, {
+    gfm: true,
+    silent: false,
+  });
+
+  // find the table
+  const rulesTable = readme.find(
+    token => token.type === 'table',
+  ) as marked.Tokens.Table;
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]: Nit - I probably should have correctly typed this

```suggestion
  ) as marked.Tokens.Table | undefined;
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  if (!rulesTable) {
+    throw Error('Could not find the rules table in README.md');
+  }
+
+  return rulesTable;
+}
+
+describe('Validating rule docs', () => {
+  it(`check if all files are present`, () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
  it('All rules must have a corresponding rule doc', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        text: `${rule.meta.docs.description} (\`${ruleName}\`)`,
+      });
+    });
+  }
+});
+
+describe('Validating rule metadata', () => {
+  for (const [ruleName, rule] of rulesData) {
+    describe(`${ruleName}`, () => {
+      it(`name field in rule must be correct`, () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
      it('`name` field in rule must match the filename', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        // validate if rule name is same as url
+        // there is no way to access this field but its used only in generation of docs url
+        expect(
+          rule.meta.docs.url.endsWith(`rules/${ruleName}.md`),
+        ).toBeTruthy();
+      });
+
+      it('Expects type checking to be set for rule', () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
      it('`requiresTypeChecking` should be set if the rule uses type information', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  }
+});
+
+describe('Validating README.md', () => {
+  const rulesTable = parseReadme().cells;
+  const notDeprecated = rulesData.filter(
+    ([, rule]) => rule.meta.deprecated !== true,
+  );
+
+  it('Checking table structure...', () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
  it('All non-deprecated rules should have a row in the table, and the table should be ordered alphabetically', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+    expect(rulesTable.map(row => row[0])).toStrictEqual(ruleNames);
+  });
+
+  for (const [ruleName, rule] of notDeprecated) {
+    describe(`Checking rule ${ruleName}`, () => {
+      const ruleRow =
+        rulesTable.find(row => row[0].includes(`/${ruleName}.md`)) ?? [];
+
+      it(`Link column should be set correctly`, () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
      it('Link column should be correct', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        expect(ruleRow[0]).toEqual(createRuleLink(ruleName));
+      });
+
+      it(`Description column should be set correctly`, () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
      it('Description column should be correct', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        expect(ruleRow[1]).toEqual(rule.meta.docs.description);
+      });
+
+      it(`Recommended column should be set correctly`, () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
      it('Recommended column should be correct', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        expect(ruleRow[2]).toEqual(
+          rule.meta.docs.recommended ? ':heavy_check_mark:' : '',
+        );
+      });
+
+      it(`Fixable column should be set correctly`, () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
      it('Fixable column should be correct', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        expect(ruleRow[3]).toEqual(
+          rule.meta.fixable !== undefined ? ':wrench:' : '',
+        );
+      });
+
+      it(`Requiring type information column should be set correctly`, () => {
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/docs.test.ts]:
```suggestion
      it('Requiring type information column should be correct', () => {
```

File: packages/eslint-plugin/tests/docs.test.ts (hunk 5 above)
Comment (armano2) [reply packages/eslint-plugin/tests/docs.test.ts]: that's correct, i just copy/paste it from tools :>

Comment (armano2) [reply packages/eslint-plugin/tests/docs.test.ts]: this actually should be:
```suggestion
const rulesTable = readme.find(
    (token): token is marked.Tokens.Table => token.type === 'table',
);
```

Comment (armano2) [reply packages/eslint-plugin/tests/docs.test.ts]: this actually should be:
```suggestion
const rulesTable = readme.find(
    (token): token is marked.Tokens.Table => token.type === 'table',
);
```

**Pull request merged**